Fix warning about unfinished oauth tasks on shutdown

twasilczyk opened the pull request:
This change adds tracking for pending OAuth tasks, so it's possible to clean them up on shutdown.

To reproduce the warning:
1. Start authentication with integration using OAuth (e.g. SmartThings)
2. When redirected to external login site, just close the page
3. Settings -> Restart Home Assistant

When shutting down HA with OAuth task pending, you'll get "Integrations should cancel non-critical tasks when receiving the stop event to prevent delaying shutdown" warning.

## Proposed change
Fix the following warning:
```
2025-03-31 22:33:02.125 WARNING (MainThread) [homeassistant.core] Task <Task pending name='Task-304' coro=<CloudOAuth2Implementation.async_generate_authorize_url.<locals>.await_tokens() running at /home/twasilczyk/projects/ha/hass-core/homeassistant/components/cloud/account_link.py:117> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[set.remove()]> was still running after final writes shutdown stage; Integrations should cancel non-critical tasks when receiving the stop event to prevent delaying shutdown
```

## Type of change
- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass** Note: on my local environment a single (test_gateway.py can't find gammu) unrelated test fails with ImportError, with or without my change.
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Evanito commented on homeassistant/components/cloud/account_link.py:
```diff
@@ -60,6 +65,27 @@ async def async_provide_implementation(
+    _PENDING_TASKS.add(task)
+
+    @callback
+    def _remove_task(_: Any) -> None:
+        _PENDING_TASKS.discard(task)
```
The task will be provided as the argument to this function. I am wondering why you chose to use this redefinition of the callback function for each task instead of using a generic callback function that could accept any task.

twasilczyk replied:
Ah, I didn't notice what's the argument type. I'll fix this and verify if it still works.

twasilczyk replied:
Removed this part of code.

Evanito reviewed (approved):
Someone should make a generic util for this type of pattern. 
`CancelableAsyncTask` or something.

MartinHjelmare commented on homeassistant/components/cloud/account_link.py:
```diff
@@ -122,12 +148,15 @@ async def await_tokens() -> None:
                 )
 
-        self.hass.async_create_task(await_tokens())
+        _async_create_task(self.hass, await_tokens())
```
I think we can simply make this a background task with `hass.async_create_background_task`.

We may need to update the test to wait for background tasks here:
https://github.com/home-assistant/core/blob/145e02769c7d58c9838178e4448a748c2e616b56/tests/components/cloud/test_account_link.py#L224

MartinHjelmare replied:
Add a comment why we need a background task, ie that we want it to cancel on Home Assistant stop, and we don't need to wait for it during setup.

twasilczyk replied:
A quick search suggests background task errors are dropped on the floor. Also, are background tasks automatically cancelled, or just hard stopped when the process exits?

Up to you, I can do it either way.

MartinHjelmare replied:
Expected exceptions in tasks should always be handled inside the task. We do that already here it looks like.

Background tasks are cancelled here when Home Assistant stops:
https://github.com/home-assistant/core/blob/597540b61120a182ad08866a161fd106bf79d2f9/homeassistant/core.py#L1140-L1144

twasilczyk replied:
Done. I still need to run full test suite on that version.

twasilczyk replied:
Tests pass except for `tests/components/sms/test_gateway.py`. I can't get `gammu` module to load.

It's unrelated anyway.

twasilczyk committed:
Don't wait for OAuth token task on shutdown

To reproduce the warning:
1. Start authentication with integration using OAuth (e.g. SmartThings)
2. When redirected to external login site, just close the page
3. Settings -> Restart Home Assistant

MartinHjelmare committed:
Clarify comment

MartinHjelmare reviewed (approved):
Thanks!
